xbmc/utils/GUIUtils.cpp

```diff
@@ -0,0 +1,363 @@
+/*
+ *  Copyright (C) 2016-2022 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "GUIUtils.h"
+
+#include "Application.h"
+#include "GUIPassword.h"
+#include "PartyModeManager.h"
+#include "PlayListPlayer.h"
+#include "ServiceBroker.h"
+#include "dialogs/GUIDialogBusy.h"
+#include "dialogs/GUIDialogKaiToast.h"
+#include "dialogs/GUIDialogYesNo.h"
+#include "filesystem/File.h"
+#include "guilib/GUIComponent.h"
+#include "guilib/GUIWindowManager.h"
+#include "guilib/LocalizeStrings.h"
+#include "messaging/helpers/DialogHelper.h"
+#include "messaging/helpers/DialogOKHelper.h"
+#include "music/tags/MusicInfoTag.h"
+#include "playlists/PlayList.h"
+#include "playlists/PlayListFactory.h"
+#include "profiles/ProfileManager.h"
+#include "settings/SettingsComponent.h"
+#include "utils/log.h"
+#include "view/GUIViewState.h"
+
+using namespace KODI::MESSAGING;
+
+using namespace XFILE;
+using namespace std::chrono_literals;
+
+using KODI::MESSAGING::HELPERS::DialogResponse;
+
+// set and reset in worker thread, read by calling thread
+CEvent CGUIUtils::m_utilsJobEvent(false, false);
+
+class CAddItemToPlaylistJob : public CJob
+{
+  const CFileItemPtr& p_Item;
+  CFileItemList& queued_Items;
+
+public:
+  CAddItemToPlaylistJob(const CFileItemPtr& pItem, CFileItemList& queuedItems)
+    : p_Item(pItem), queued_Items(queuedItems)
+  {
+  }
+
+  ~CAddItemToPlaylistJob(void) override = default;
+
+  bool DoWork(void) override
+  {
+    CGUIUtils sharedUtils;
+    sharedUtils.StartQueue();
+    sharedUtils.AddToPlayList(p_Item, queued_Items);
+    sharedUtils.QueueComplete(); // Signal job finished
+    return true;
+  }
+};
+
+bool CGUIUtils::PlayItem(const CFileItemPtr& itemToPlay)
+{
+  if (itemToPlay->m_bIsFolder) // build a playlist and play it
+  {
+    XFILE::CMusicDatabaseDirectory dir;
+    // Don't add top level nodes to the playlist - should be excluded by the calling code already
+    XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE Node =
+        dir.GetDirectoryParentType(itemToPlay->GetPath());
+    if (Node == XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE_OVERVIEW)
+      return false;
+    CFileItemList queuedItems;
+    AddItemToPlayList(itemToPlay, queuedItems);
+    CServiceBroker::GetPlaylistPlayer().ClearPlaylist(PLAYLIST_MUSIC);
+    CServiceBroker::GetPlaylistPlayer().Reset();
+    CServiceBroker::GetPlaylistPlayer().Add(PLAYLIST_MUSIC, queuedItems);
+    CServiceBroker::GetPlaylistPlayer().SetCurrentPlaylist(PLAYLIST_MUSIC);
+    CServiceBroker::GetPlaylistPlayer().Play();
+  }
+  else // song, so just play it
+  {
+    CFileItem item(itemToPlay->GetItemToPlay());
+    PlaySong(item);
+  }
+  m_musicDatabase.Close();
```

Comment:
Where is it opened, why must it be open in this method? Why can it be closed at the end of this method? This needs explanation if not a fix.